src/edgepi/dac/edgepi_dac.py

```diff
@@ -195,59 +195,86 @@ def compute_expected_voltage(self, analog_out: DACChannel) -> float:
         Return:
             code (int): modified code value
         """
-        if enable:
+        if toggle_gain:
             return int(code/CalibConst.DAC_GAIN_FACTOR.value)
-        if code<CalibConst.RANGE.value/2:
-            return code*CalibConst.DAC_GAIN_FACTOR.value
+
+        code = code*CalibConst.DAC_GAIN_FACTOR.value
+        if code > CalibConst.RANGE.value:
+            raise ValueError(f"Code Value: {code} is out of range")
 
         return code
 
-    def __get_ch_codes(self, enable: bool = None):
+    def __get_ch_codes(self, toggle_gain: bool = None):
         """
         Read and modify channel code value
         Args:
-            enable(bool): False: multiply the current code value by 2
+            toggle_gain (bool): False: multiply the current code value by 2
                           True: divide the current code value by 2
         Return:
             code_vals (list): list of code values from ch1-ch8
         """
         code_vals = []
         for ch in DACChannel:
             code, _, _ = self.get_state(ch, True, False, False)
-            code_vals.append(self.__modify_code_val(enable, code))
+            code_vals.append(self.__compute_code_val(toggle_gain, code))
         return code_vals
 
-    def enable_dac_gain(self, enable: bool = None, ch_code_handler: bool = False):
+    def __auto_code_handler(self, toggle_gain: bool = None):
         """
-        Enable/Disable internal DAC gain by toggling the DAC_GAIN pin
+        Compute and change code values of each channel when toggling DAC gain
         Args:
-            enable (bool): enable boolean to set or clear the gpio pin
-            ch_code_handler (bool): flag to re-write code value of each channel to keep the same
-                                    output voltage
+            toggle_gain (bool): enable boolean to set or clear the gpio pin
         Return:
-            gain_state (bool): state of the gain pin
+            N/A
         """
-        if ch_code_handler and self.__get_gain_state() != enable:
-            codes = self.__get_ch_codes(enable)
-            self.log.debug(f'Code: {codes}')
+        codes = self.__get_ch_codes(toggle_gain)
+        self.log.debug(f'Code: {codes}')
+
+        # gain being enabled, change code first than enable gain
+        if toggle_gain:
             for ch, code in enumerate(codes):
                 # update DAC register
                 self.transfer(self.dac_ops.generate_write_and_update_command(ch, code))
+            self.gpio.set_pin_state(GainPin.DAC_GAIN.value)
 
+        # Disabling gain, change gain first than change codes
+        else:
+            self.gpio.clear_pin_state(GainPin.DAC_GAIN.value)
+            for ch, code in enumerate(codes):
+                # update DAC register
+                self.transfer(self.dac_ops.generate_write_and_update_command(ch, code))
+
+    def toggle_dac_gain(self, toggle_gain: bool = None, auto_code_change: bool = False):
```

Comment:
toggle is not a good name either. toggle is usually used when you want to change the state of something to the opposite of what it is currently. here we want to set gain to something specific. how about:
```
def set_dac_gain(self, is_gain_enabled: bool, auto_code_adjust: bool = False):
```
also I don't think we should have a default value for gain.